contracts/external/cw-payroll/src/contract.rs

```diff
@@ -53,13 +60,105 @@ pub fn execute(
         ExecuteMsg::Receive(msg) => execute_receive(env, deps, info, msg),
         ExecuteMsg::Distribute { id } => execute_distribute(env, deps, info, id),
+        ExecuteMsg::PauseStream { id } => execute_pause_stream(env, deps, info, id),
+        ExecuteMsg::ResumeStream {
+            id,
+            start_time,
+            end_time,
+        } => execute_resume_stream(env, deps, info, id, start_time, end_time),
+        ExecuteMsg::RemoveStream { id } => execute_remove_stream(env, deps, info, id),
     }
 }
+pub fn execute_pause_stream(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: StreamId,
+) -> Result<Response, ContractError> {
+    let mut stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound {})?;
+
+    if stream.admin != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+    if stream.paused {
+        return Err(ContractError::StreamAlreadyPaused {});
+    }
+    stream.paused_time = Some(env.block.time.seconds());
+    stream.paused = true;
+    save_stream(deps, id, &stream).unwrap();
+    let response = Response::new()
+        .add_attribute("method", "pause_stream")
+        .add_attribute("paused", stream.paused.to_string())
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", info.sender)
+        .add_attribute("paused_time", env.block.time.to_string());
 
+    Ok(response)
+}
+pub fn execute_remove_stream(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: StreamId,
+) -> Result<Response, ContractError> {
+    let stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound {})?;
+
+    if stream.admin != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    remove_stream(deps, id).unwrap();
+    let response = Response::new()
+        .add_attribute("method", "remove_stream")
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", info.sender)
+        .add_attribute("paused_time", env.block.time.to_string());
+
+    Ok(response)
+}
+pub fn execute_resume_stream(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: StreamId,
+    _start_time: Option<u64>,
+    _end_time: Option<u64>,
+) -> Result<Response, ContractError> {
+    let mut stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound {})?;
+
+    if stream.admin != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+    if !stream.paused {
+        return Err(ContractError::StreamNotPaused {});
+    }
+    let duration: u128 = (stream.end_time - stream.start_time).into();
+    let paused_duration: u128 = (env.block.time.seconds() - stream.paused_time.unwrap()).into();
+    let rate_per_second = Uint128::from(stream.balance.total_amount() / duration - paused_duration);
+    stream.paused = false;
+    stream.rate_per_second = rate_per_second;
```

Comment:
See comment in msg or state -- I don't think we should store `rate_per_second`, can always be computed on request.
So, remove `stream.rate_per_second` but you can leave inside response attribute

Reply:
But when pause happened multiple times, we should recalculate all this pauses, so we need store this value